pkg/module/memfs.go

```diff
@@ -0,0 +1,43 @@
+package module
+
+import (
+	"io"
+	"io/fs"
+	"path/filepath"
+
+	"golang.org/x/xerrors"
+
+	dio "github.com/aquasecurity/go-dep-parser/pkg/io"
+	"github.com/aquasecurity/memoryfs"
+)
+
+// memFS is a wrapper of memoryfs.FS and can change its underlying file system
+// at runtime. This implements fs.FS.
+type memFS struct {
+	current *memoryfs.FS
+}
+
+// Open implements fs.FS.
+func (m *memFS) Open(name string) (fs.File, error) {
+	return m.current.Open(name)
+}
+
+// initialize changes the underlying memory file system with the given file path and contents.
+//
+// Note: it is always to safe swap the underlying FS with this API since this is called only at the beginning of
```

Comment:
The same module could be called concurrently. For example, Trivy walks filesystem and calls `Analyze` on different files in parallel.

- /dir
  - a.txt: call Analyze()
  - b.txt: call Analyze()

Seems like the above case is problematic if I understand correctly.

Reply:
Oh then the current implementation is already problematic and there’s a race regardless of thisPR, isn’t it?

Reply:
Wash module in general is not thread safe

Reply:
So general approach to this Wasm concurrency is 1) instantiate-per-request and 2) the pool of instances, and 2) is favored by variety of projects such as mosn, http-wasm. cc @codefromthecrypt

Reply:
indeed, filesystem would only be one issue if something goes in parallel. For example, known garbage collection implementations are not implemented in a thread-safe way by compilers. Since most end user code ends up allocating at least strings, this can lead to issues.

We have code to re-use which way out, so let us know which way below you prefer @knqyf263!

1. lock single instance on use (same code, just locked in the critical sections like analyze)
  * even if things are parallel, it may not reflect the outcome too much depending on how fast analyze returns 
2. pooled instances
  * may also need a limiter semaphore if that's not already handled upstream in the correct way
3. instantiate-per-request
  * if this is one per file analyzed, maybe the overhead will be too much.

Reply:
>Oh then the current implementation is already problematic and there’s a race regardless of thisPR, isn’t it?

In terms of the filesystem, `memfs` is initialized in `Analyze` and it should be thread-safe now.
https://github.com/aquasecurity/trivy/blob/5c434753ce25c65c53c5310b264cd994062aa2ee/pkg/module/module.go#L420

However, if `WithFS` anyway swaps the underlying filesystem and doesn't lock it (it is the case AFAIR) , the current implemnetation is not thread-safe regardless of this PR like you said...
https://github.com/aquasecurity/trivy/blob/5c434753ce25c65c53c5310b264cd994062aa2ee/pkg/module/module.go#L432

>1. lock single instance on use (same code, just locked in the critical sections like analyze)

From the perspective of the performance, I prefer 2, but it is experimental. We can go for 1.

Reply:
even memfs is safe, but `m.analyze.Call` is called concurrently? 

https://github.com/aquasecurity/trivy/blob/5c434753ce25c65c53c5310b264cd994062aa2ee/pkg/module/module.go#L441

Reply:
that's totally unsafe as @codefromthecrypt said

Reply:
good, will do the 1. first !

Reply:
https://github.com/aquasecurity/trivy/pull/3299/commits/638aa32d1ba6f2ce1ff730b970e7745796d78103 done!